runtime/core/frontend/feature_pipeline.cc

```diff
@@ -92,23 +95,61 @@ bool FeaturePipeline::ReadOne(std::vector<float>* feat) {
 
 bool FeaturePipeline::Read(int num_frames,
                            std::vector<std::vector<float>>* feats) {
+  bool b = true;
   feats->clear();
-  std::vector<float> feat;
-  while (feats->size() < num_frames) {
-    if (ReadOne(&feat)) {
-      feats->push_back(std::move(feat));
+  std::vector<std::vector<float>> tmp_feats;
+  std::vector<std::vector<float>> chunk_feats;
+  tmp_feats.insert(tmp_feats.end(), remained_feats_.begin(),
+                   remained_feats_.end());
+  while (tmp_feats.size() < num_frames) {
+    if (ReadFromQueue(&chunk_feats)) {
+      tmp_feats.insert(tmp_feats.end(), chunk_feats.begin(), chunk_feats.end());
     } else {
-      return false;
+      b = false;
+      break;
     }
   }
-  return true;
+  if (num_frames > tmp_feats.size()) {
+    num_frames = tmp_feats.size();
+  }
+  if (num_frames > 0) {
+    feats->insert(feats->end(), tmp_feats.begin(),
+                  tmp_feats.begin() + num_frames);
+    remained_feats_.resize(tmp_feats.size() - num_frames);
+    std::copy(tmp_feats.begin() + num_frames, tmp_feats.end(),
+              remained_feats_.begin());
+    {
+      std::lock_guard<std::mutex> lock(wait_mutex_);
+      wait_frames_ -= num_frames;
+    }
+  }
+  return b;
+}
+
+int FeaturePipeline::NumQueuedFrames() const {
+  std::lock_guard<std::mutex> lock(wait_mutex_);
+  return wait_frames_;
 }
 
 void FeaturePipeline::Reset() {
   input_finished_ = false;
   num_frames_ = 0;
   remained_wav_.clear();
+  remained_feats_.clear();
   feature_queue_.Clear();
+  {
+    std::lock_guard<std::mutex> lock(wait_mutex_);
+    wait_frames_ = 0;
+  }
+}
+bool FeaturePipeline::ReadOne(std::vector<float>* feat) {
```

Comment:
Do we still need this function?

Reply:
I have considered compatibility. It was originally a public function. I searched all the codes of wenet and found no calls. However, I am not sure whether the developers based on wenet have called this function.